src/scanner.c

```diff
@@ -0,0 +1,465 @@
+#include <tree_sitter/parser.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <wctype.h>
+
+typedef enum
+{
+  LINE_START,
+  LINE_CONTENT,
+  DASHES,
+  AFTER_DASHES,
+} MatchState;
+
+typedef enum
+{
+  NEWLINE,
+  INDENT,
+  DEDENT,
+  STRING_START,
+  STRING_CONTENT,
+  STRING_END,
+  COMMENT,
+} TokenType;
+
+typedef struct
+{
+  char flags;
+} Delimiter;
+
+#define SINGLE_QUOTE (1 << 0)
+#define DOUBLE_QUOTE (1 << 1)
+
+typedef struct
+{
+  uint16_t previous_indent_length;
+  Delimiter *delimiter_stack;
+  size_t delimiter_stack_size;
+  size_t delimiter_stack_capacity;
+} Scanner;
+
+static void Delimiter_init(Delimiter *delimiter)
+{
+  delimiter->flags = 0;
+}
+
+static int32_t Delimiter_end_character(const Delimiter *delimiter)
+{
+  if (delimiter->flags & SINGLE_QUOTE)
+    return '\'';
+  if (delimiter->flags & DOUBLE_QUOTE)
+    return '"';
+  return 0;
+}
+
+static void Delimiter_set_end_character(Delimiter *delimiter, int32_t character)
+{
+  switch (character)
+  {
+  case '\'':
+    delimiter->flags |= SINGLE_QUOTE;
+    break;
+  case '"':
+    delimiter->flags |= DOUBLE_QUOTE;
+    break;
+  default:
+    assert(false);
+  }
+}
+
+static Scanner *Scanner_new()
+{
+  Scanner *scanner = (Scanner *)malloc(sizeof(Scanner));
+  assert(scanner != NULL);
+  scanner->previous_indent_length = 0;
+  scanner->delimiter_stack = NULL;
+  scanner->delimiter_stack_size = 0;
+  scanner->delimiter_stack_capacity = 0;
```

Comment:
notice how it has inlined a basic vector implementation. It's worth a quick look at how it implements this `delimiter_stack` stuff to make sure it's not doing anything weird